src/imperative/pointwise_fusion_pass.cc

```diff
@@ -57,281 +58,321 @@ void WarnFusionNotSupported() {
+                               output.version);
+    } else {
+      const int output_num = SetInsert({static_cast<int>(output.node_id),
+                                        static_cast<int>(output.index)},
+                                       &(subgraphs[subgraph_id].outputs));
+      ret.outputs.emplace_back(subgraphs[subgraph_id].subgraph_node,
+                               output_num,
+                               output.version);
     }
   }
-  std::vector<std::vector<nnvm::Node*> > to_add(subsets->size());
-  DFSVisit(g.outputs, [&is_compatible, &node2setidx, &to_add](const nnvm::ObjectPtr& n) {
-    const auto& it = node2setidx.find(n.get());
-    if (it != node2setidx.end()) {
-      for (auto& e : n->inputs) {
-        if (is_compatible(e.node.get()))
-          to_add[it->second].push_back(e.node.get());
-      }
+
+  for (auto& info : subgraphs) {
+    info.graph.outputs.reserve(info.outputs.size());
+    for (const auto& entry_info : info.outputs) {
+      info.graph.outputs.emplace_back(new_nodes[entry_info.source_node],
+                                      entry_info.index,
+                                      0);
     }
-  });
+    create_subgraph_node(info.graph, info.inputs.size(), info.subgraph_node.get());
+  }
 
-  // Avoid duplicating the node that is input of two subsets
-  std::unordered_set<nnvm::Node*> added;
-  for (size_t i = 0; i < subsets->size(); ++i) {
-    std::vector<nnvm::NodeEntry> heads;
-    for (auto n : subsets->at(i)) {
-      for (auto e : n->inputs) {
-        if (!subsets->at(i).count(e.node.get()))
-          heads.push_back(e);
+  for (size_t i = 0; i < idx.num_nodes(); ++i) {
+    // Add _FusedOpHelper nodes
+    const int subgraph_id = subgraph_assignment[i];
+    for (size_t dep_num = 0; dep_num < idx[i].control_deps.size(); ++dep_num) {
+      const auto& dep = idx[i].control_deps[dep_num];
+      const int their_subgraph_id = subgraph_assignment[dep];
+      if (subgraph_assignment[i] != -1 && subgraph_assignment[dep] == -1) {
+        // Not in any subgraph, use FusedOpOutHelper
+        auto& info = subgraphs[subgraph_id];
+        size_t node_id = info.subgraph_node->control_deps.size();
+        info.subgraph_node->control_deps.emplace_back(new_nodes[dep]);
+        auto helper_node = op::MakeNode("_FusedOpOutHelper",
+                                        "FusedOp_" + new_nodes[i]->attrs.name + "_outhelper",
+                                        nullptr,
+                                        nullptr,
+                                        nullptr);
+        helper_node->attrs.parsed =
+          FusedOpHelperParamPtr(new FusedOpHelperParam(
+                nnvm::get<FusedOpPtr>(info.subgraph_node->attrs.parsed),
+                node_id));
+        new_nodes[i]->control_deps.insert(new_nodes[i]->control_deps.begin() + dep_num,
+                                          std::move(helper_node));
+      } else if (their_subgraph_id != subgraph_id &&
```

Comment:
can `subgraph_id` be -1 here? or should we add that check here?

Reply:
It can. The role of `FusedOpHelper` is to get information (for shape/type inference) from inside the `FusedOp` to the outside world.

Reply:
So if your dependency is going to be fused, then you need to have this helper instead.

Reply:
Makes sense. In general, what are the control flow dependencies of a node? I know of only one - dependency of backward op on its corresponding forward op.

Reply:
Yeah, that is the dependency we are changing as it is used for infershape/infertype. Going forward I would like to get rid of this and never do the inferattr on the fused graph (instead do it on the original graph and map the results to the fused one), but it is outside the scope of this PR.